### PR TITLE
【paddle.fleet】Paddle fleet distributed strategy

### PR DESCRIPTION
**Pull request by guru4elephant**
### PR types

New features

### PR changes

APIs

### Describe

unify distributed_strategy in paddle.fleet. All distributed training strategy should be configured through this class.

Propose to formalize the paddle.fleet api as follows:

base
- distributed_strategy.py
  - all distributed strategies are implemented here, including various kinds of distributed optimizers
  - the distributed strategy can be serialized into protobuf file
  - distributed job info is defined in distributed_strategy.py, including distributed runtime information such as endpoints, node num, etc.
  - a distributed training job can be configured through serialized distributed strategy and distributed job info.
- fleet_base.py
  - the only entry of distributed training of Paddle singleton instance
a Fleet() object is initialized through RoleMaker
Util and DistributedOptimizer(either collective、ps or other distributed optimizer) can be initialized by RoleMaker
obj_creator.py
obj_creator helps to create Util and DistributedOptimizer
- role_maker.py
  - PaddleCloudRoleMaker and UserDefinedRoleMaker can be defined currently different kinds of RoleMaker is specified by arguments of __init__(self, role_maker)
- util_base.py
  - implement all Util classes, Util Base class can be inherited so that a third party Util class can be used.

collective
- collective_distributed_optimizer.py
  - define collective training
  - parameter_server
- ps_distributed_optimizer.py
  - define ps training

metrics
- metric.py
  - distributed metrics

What is paddle.fleet?
- paddle.fleet is the official entry point for distributed training of Paddle. The motivation of paddle.fleet is to make distributed training efficient and easy to configure.
- RoleMaker, DistributedStrategy and DistributedOptimizer are the key concepts to understand in paddle.fleet
  - RoleMaker : the role of a node in distributed cluster is specified by RoleMaker, such as worker or pserver.
  - DistributedStrategy : paddle.fleet supports lots of distributed training algorithm and scalable configurations all defined in DistributedStrategy, such as hierarchical all reduce, deep gradient compression and async parameter server training, etc.
  - DistributedOptimizer : in general, three kinds of optimizers exist in paddle. 1) commonly used local optimizer, such as SGD and Adam. 2) decorator of local optimzier, such RecomputeOptimizer, PipelineOptimizer and LookAheadOptimizer that often wraps a local optimizer as an inner optimizer. 3) distributed decorator of 1)、2) or combination of 1) and different 2), such as collective distributed optimizer and parameter server distributed optimzier. paddle.fleet is the third kind of optimizer that users can define.

### CODE REVIEW

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by swtkiwi — APPROVED** [review]
LGTM

**Review by gavin1332 — APPROVED** [review]
LGTM

File: python/paddle/fleet/base/distributed_strategy.py [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+            print("WARNING: sync should have value of bool type")
+
+    @property
+    def async(self):
+        return self.strategy.async
+
+    @async.setter
+    def async(self, flag):
+        if isinstance(flag, bool):
+            self.strategy.async = flag
```
Comment (raindrops2sea) [inline python/paddle/fleet/base/distributed_strategy.py:384]: 如果asyn和sync的flag被设置为相同，比如都为true，怎么办？

File: python/paddle/fleet/__init__.py [hunk 2]
```diff
@@ -13,16 +13,11 @@
...
-#            '	DistributeTranspilerConfig',
-#            '	HashName',
-#            '	RoundRobin',
-#            'collective']
+from .base.distributed_strategy import DistributedStrategy
+#from .base.role_maker import PaddleCloudRoleMaker, UserDefinedRoleMaker
+#from .base.fleet_base import Fleet
+
+#__all__ = [
+#    "DistributedStrategy", "PaddleCloudRoleMaker", "UserDefinedRoleMaker"
```
Comment (raindrops2sea) [inline python/paddle/fleet/__init__.py:21]: RoleMaker等下一个pr再添加？

**Review by raindrops2sea — APPROVED** [review]
LGTM

File: python/paddle/fleet/__init__.py (hunk 2 above)
Comment (guru4elephant) [reply python/paddle/fleet/__init__.py:21]: 是的，占位

File: python/paddle/fleet/base/distributed_strategy.py (hunk 1 above)
Comment (guru4elephant) [reply python/paddle/fleet/base/distributed_strategy.py:384]: 互斥的情况下，按strategy优先级生效

**Pull request merged**